### PR TITLE
[iOS] Fabric: Fixes custom font of weight is not honored

### LINKED ISSUE #47656: Font weight is not honored on iOS in RN 0.76
### Description

After switching to new architecture, the text rendering does not seem to honor the `fontWeight` style as it used to do, on iOS.

With RN 0.76 and new architecture enabled, it's not choosing the font based on the weight on iOS, even though that's the intended behavior according to the docs: https://reactnative.dev/docs/text-style-props#fontweight

### Steps to reproduce

clone the project and run 
`yarn && npx pod-install && yarn ios`

Or add this code to an existing RN 0.76 project with new architecture enabled:

```jsx
<Text
  style={{
    fontFamily: 'RobotoCondensed-Bold',
    fontSize: 32,
    fontWeight: '700',
  }}>
    RobotoCondensed-Bold - weight 700
</Text>
<Text
    style={{
    fontFamily: 'RobotoCondensed-Regular',
    fontSize: 32,
    fontWeight: '700',
  }}>
    RobotoCondensed-Regular - weight 700
</Text>
```

### React Native Version

0.76.2

### Affected Platforms

Runtime - iOS

### Areas

Fabric - The New Renderer

### Output of `npx react-native info`

```text
System:
  OS: macOS 15.0.1
  CPU: (10) arm64 Apple M1 Max
  Memory: 819.03 MB / 64.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 18.18.2
    path: ~/Library/Caches/fnm_multishells/40214_1731916352195/bin/node
  Yarn:
    version: 1.22.22
    path: ~/Library/Caches/fnm_multishells/40214_1731916352195/bin/yarn
  npm:
    version: 9.8.1
    path: ~/Library/Caches/fnm_multishells/40214_1731916352195/bin/npm
  Watchman:
    version: 2024.10.07.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.15.2
    path: /opt/homebrew/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.5
      - iOS 17.5
      - macOS 14.5
      - tvOS 17.5
      - visionOS 1.2
      - watchOS 10.5
  Android SDK:
    API Levels:
      - "27"
      - "28"
      - "29"
      - "31"
      - "32"
      - "33"
      - "34"
      - "34"
      - "35"
    Build Tools:
      - 28.0.3
      - 29.0.2
      - 30.0.2
      - 30.0.3
      - 31.0.0
      - 33.0.0
      - 34.0.0
      - 35.0.0
    System Images:
      - android-28 | Google ARM64-V8a Play ARM 64 v8a
      - android-29 | Intel x86 Atom_64
      - android-29 | Google APIs Intel x86 Atom
      - android-33 | Google APIs ARM 64 v8a
      - android-34 | Google Play ARM 64 v8a
    Android NDK: 25.0.8775105
IDEs:
  Android Studio: 2023.1 AI-231.9392.1.2311.11255304
  Xcode:
    version: 15.4/15F31d
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 1.8.0_292
    path: /usr/bin/javac
  Ruby:
    version: 3.3.5
    path: /opt/homebrew/opt/ruby/bin/ruby
npmPackages:
  "@react-native-community/cli":
    installed: 15.0.1
    wanted: 15.0.1
  react:
    installed: 18.3.1
    wanted: 18.3.1
  react-native:
    installed: 0.76.2
    wanted: 0.76.2
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: true
iOS:
  hermesEnabled: true
  newArchEnabled: true
```

### Stacktrace or Logs

```text
N/A
```

### Reproducer

https://github.com/esbenvb/rn-font-weight-issue

### Screenshots and Videos

✅ Without new architecture - working as expected
Simulator Screenshot - iPhone 15 - 2024-11-18 at 09 00 58

❌ With new architecture - not working as expected 
Simulator Screenshot - iPhone 15 - 2024-11-18 at 09 03 57

### PR DESCRIPTION
**Pull request by zhongwuzw**
## Summary:

Fixes #47656 .

## Changelog:

[IOS] [FIXED] - Fabric: Fixes custom font of weight is not honored

## Test Plan:

Demo in #47656

### CODE REVIEW

File: packages/react-native/ReactCommon/react/renderer/textlayoutmanager/platform/ios/react/renderer/textlayoutmanager/RCTFontUtils.mm [hunk 1]
```diff
@@ -166,6 +166,13 @@ static RCTFontStyle RCTGetFontStyle(UIFont *font)
   } else {
     NSArray<NSString *> *fontNames = [UIFont fontNamesForFamilyName:fontProperties.family];
 
+    if (fontNames.count == 0) {
+      font = [UIFont fontWithName:fontProperties.family size:effectiveFontSize];
+      if (font) {
+        fontNames = [UIFont fontNamesForFamilyName:font.familyName];
+      }
+    }
+
```
Comment (javache) [inline packages/react-native/ReactCommon/react/renderer/textlayoutmanager/platform/ios/react/renderer/textlayoutmanager/RCTFontUtils.mm:175]: This logic is very similar to what's immediately below - should it be restructured more? 

```
fontNames = [...]
if (fontNames.count == 0) {
  font = [UIFont fontWithName:fontProperties.family size:effectiveFontSize];
  if (font) {
    fontNames = [UIFont fontNamesForFamilyName:font.familyName];
  } else {
    // Fallback to system font.
    font = [UIFont systemFontOfSize:effectiveFontSize weight:fontProperties.weight];
}

if (fontNames.count >  1) {}

**Review by javache — COMMENTED** [review]
Can you explain the fix here? Why does `fontWithName:size` while `fontNamesForFamilyName` does not?

**Comment by zhongwuzw** [comment]
> Can you explain the fix here? Why does `fontWithName:size` while `fontNamesForFamilyName` does not?

@javache The reason please see the comment as below: 

```
// Gracefully handle being given a font name rather than font family, for
// example: "Helvetica Light Oblique" rather than just "Helvetica".
```

**Comment by esbenvb** [comment]
Does not work with Chivo font, it renders Regular as Bold... 

image

**Comment by esbenvb** [comment]
Thanks for the quick attempt for a fix!

I can see this one has now been merged, but as mentioned in #47656  it doesn't solve all the issues in the, and the code/fonts that has been working with the old architecture and still works on Android, doesn't look right now.

What can be done from here? Ideally, we'd like it to behave just like before the upgrade to RN 0.76

**Comment by zhongwuzw** [comment]
> Thanks for the quick attempt for a fix!
> 
> 
> 
> I can see this one has now been merged, but as mentioned in #47656  it doesn't solve all the issues in the, and the code/fonts that has been working with the old architecture and still works on Android, doesn't look right now.
> 
> 
> 
> What can be done from here? Ideally, we'd like it to behave just like before the upgrade to RN 0.76

Hi, I will try to investigate other issues of font you provided.

**Pull request closed without merging**